src/type_conversion/hiwire.c

```diff
@@ -2,51 +2,115 @@
 
 #include "hiwire.h"
 
+int
+hiwire_error()
+{
+  return HW_ERROR;
+}
+
+int
+hiwire_undefined()
+{
+  return HW_UNDEFINED;
+}
+
+int
+hiwire_null()
+{
+  return HW_NULL;
+}
+
+int
+hiwire_true()
+{
+  return HW_TRUE;
+}
+
+int
+hiwire_false()
+{
+  return HW_FALSE;
+}
+
+int
+hiwire_bool(int boolean)
+{
+  return boolean ? hiwire_true() : hiwire_false();
+}
+
 EM_JS(void, hiwire_setup, (), {
-  // These ids must match the constants above, but we can't use them from JS
-  var hiwire = { objects : {}, counter : 1 };
-  hiwire.objects[-2] = undefined;
-  hiwire.objects[-3] = true;
-  hiwire.objects[-4] = false;
-  hiwire.objects[-5] = null;
-
-  Module.hiwire_new_value = function(jsval)
+  let _hiwire = { objects : new Map(), counter : 1 };
+  Module.hiwire = {};
+  Module.hiwire.ERROR = _hiwire_error();
+  Module.hiwire.UNDEFINED = _hiwire_undefined();
+  Module.hiwire.NULL = _hiwire_null();
+  Module.hiwire.TRUE = _hiwire_true();
+  Module.hiwire.FALSE = _hiwire_false();
+
+  _hiwire.objects.set(Module.hiwire.UNDEFINED, undefined);
+  _hiwire.objects.set(Module.hiwire.NULL, null);
+  _hiwire.objects.set(Module.hiwire.TRUE, true);
+  _hiwire.objects.set(Module.hiwire.FALSE, false);
+
+  Module.hiwire.new_value = function(jsval)
   {
-    var objects = hiwire.objects;
-    while (hiwire.counter in objects) {
-      hiwire.counter = (hiwire.counter + 1) & 0x7fffffff;
+    // Should we guard against duplicating standard values?
+    // Probably not worth it for performance: it's harmless to ocassionally
+    // duplicate. Maybe in test builds these should raise? if(jsval ===
+    // undefined){
+    //   return Module.hiwire.UNDEFINED;
+    // }
+    // if(jsval === null){
+    //   return Module.hiwire.NULL;
+    // }
+    // if(jsval === true){
+    //   return Module.hiwire.TRUE;
+    // }
+    // if(jsval === false){
+    //   return Module.hiwire.FALSE;
+    // }
```

Comment:
Please remove commented code, unless it's part of the comment?

Reply:
Okay will remove.